designs/2020-generic-ast-support/README.md

```diff
@@ -0,0 +1,57 @@
+
+## Detailed Design
+
+Currently ESLint has a lot of hardcoded assumptions about AST. For example, it assumes that nodes are identified by the property called `type` and that property exists on every node. It assumes that top-most node is always `Program` and that AST supports code-path-analysis (which is very specific to JavaScript). This RFC proposes to modify expected output from the `parseForESLint` method of the custom parsers to include additional property called `parserSupportOptions` of type object with four new properties (for now):
+ - nodeIdentifier: string - Specifies name of the ASTNode property that identifies node. Defaults to `type` for backwards compatibility.
+ - supportCodePathAnalysis: Boolean - Indicates AST's support for built-in CodePathAnalysis support. Defaults to `true` for backwards compatibility.
+ - supportScope: Boolean - Indicates AST's support for scope analysis. Defaults to `true` for backwards compatibility.
+ - rootNode: string - Provides the name of the top level node in the AST. Defaults to `Program` for backwards compatibility.
+
+ This would change would allow basic support of linting non-estree compliant ASTs. This will still not enable full feature parity with ESLint's abilities when linting JavaScript. In the future new properties might be added to support parsing comments to enable inline directives for controlling rules. Other enhancements might include ability to provide replacement for CodePathAnalysis and scopes creation.
+
+## Documentation
+
+This will be documented as part of the page that describes [working with custom parsers](https://eslint.org/docs/developer-guide/working-with-custom-parsers). In addition, blog post would let other users know that the can start creating parsers and rules for new languages. Although, it might be a good idea to wait until there's an example in the wild that could be pointed to as a reference.
+
+## Drawbacks
+
+The only drawback that I can think of, is that JavaScript specific rules will not work for other ASTs. Each language will most likely have to implement their own set of rules, and they will be incompatible across different parsers. While this might cause some confusion for users, I do believe this makes sense and there is no need to try to somehow generalize rules to support multiple ASTs.
+
+## Backwards Compatibility Analysis
+
+This change will be fully backwards compatible. In the future, default described above could be removed and moved into espree instead.
+
+## Alternatives
+
+None that I can think of. Other then not to support any ASTs other then estree compliant.
+
+## Open Questions
+
+* Is it OK to release incomplete support for other AST types? Inline directives being the biggest obstacle, from my perspective.
```

Comment:
Yes, and I think that's actually preferable. It's my impression that we all sort of stumbled onto the realization around the same time that ESLint is closer to supporting non-ESTree ASTs than we might have thought before investigating. If we keep this as experimental, void-your-warranty type stuff, then we get to see where people want to take it rather than building the "perfect" design in an ivory tower before we know what the requirements will be.